DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,235 @@
+        }
+
+        #region IEnumerable Implementation
+        public IEnumerator<T> GetEnumerator() 
+        {
+            for (int i = _start; i < Count; i++) 
+            {
+                yield return _circularBuffer[i];
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() 
+        {
+            return GetEnumerator();
+        }
+        #endregion
+
+        #region ICollection Implementation
+        private int _count;
+        /// <summary>
+        /// Returns the number of elements.
+        /// </summary>
+        public int Count 
+        {
+            get 
+            {
+                return _count;
+            }
+        }
+        /// <summary>
+        /// Checks whether this collection is readonly
+        /// </summary>
+        public bool IsReadOnly 
+        {
+            get 
+            {
+                return false;
+            }
+        }
+        /// <summary>
+        /// Clears this instance
+        /// </summary>
+        public void Clear() 
+        {
+            _count = 0;
+            _start = 0;
+            _end = 0;
+            _circularBuffer = new T[Length + 1];
+        }
+        /// <summary>
+        /// Checks whether the buffer contains an item
+        /// </summary>
+        public bool Contains(T item) 
+        {
+            return _circularBuffer.Contains(item);
+        }
+        /// <summary>
+        /// Copies this buffer to an array
+        /// </summary>
+        public void CopyTo(T[] array, int arrayIndex) 
+        {
+            if (array == null) 
+            {
+                throw new ArgumentNullException("array can not be null");
+            }
+
+            if (array.Length == 0 || arrayIndex >= array.Length || arrayIndex < 0) 
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            // Get enumerator
+            var enumarator = GetEnumerator();
+
+            // Copy elements if there is any in the buffer and if the index is within the valid range
+            while (arrayIndex < array.Length) 
+            {
+                if (enumarator.MoveNext()) 
+                {
+                    array[arrayIndex] = enumarator.Current;
+                    arrayIndex++;
+                }
+                else 
+                {
+                    break;
+                }
+            }
+        }
+        /// <summary>
+        /// Removes an item from the buffer
+        /// </summary>
+        public bool Remove(T item) 
+        {
+            if (!IsEmpty && Contains(item)) 
+            {
+               var sourceArray = _circularBuffer.Except(new T[] { item }).ToArray();
+                _circularBuffer = new T[Length + 1];
+                Array.Copy(sourceArray, _circularBuffer, sourceArray.Length);
+
+                if (!item.Equals(default(T))) 
+                {
+                    _end = sourceArray.Length - 1;
+                    _count = sourceArray.Length-1;
```

Comment:
`_count` calculation is wrong. Count *IS* the length of the `sourceArray`, since that contains the remaining elements.

Reply:
Apologies for not replying earlier.
The `_count` calculation is not wrong. Remember that for the case where `item` is not equal to `default(T)`, the `sourceArray` will be reduced to `{[values], <empty>}`. So, the `_count` is the length of the [values] minus the <empty> space.

Reply:
Yes, it is wrong when we are removing `default(T)`, because the length of the `sourceArray` exactly matches the number of elements left. But, it is correct when we are removing other values, because the `sourceArray` will contain `{[other values],<empty>}`. Therefore the count should be one less than the length of `sourceArray`.